src/python/WMCore/Storage/RucioFileCatalog.py

```diff
@@ -0,0 +1,236 @@
+    instance containing the details found in it
+    """
+    rfcInstance = RucioFileCatalog()
+    try:
+        jsonFile = open(filename,encoding="utf-8")
+        jsElements = json.load(jsonFile)
+    except Exception as ex:
+        msg = "Error reading storage description file: %s\n" % filename
+        msg += str(ex)
+        raise RuntimeError(msg)
+    #now loop over elements, select the right one and fill lfn-to-pfn
+    for jsElement in jsElements:
+        #check to see if the storageSite and volume matchs with "site" and "volume" in storage.json
+        if jsElement['site'] == storageSite and jsElement['volume'] == volume: 
+            rfcInstance.preferredProtocol = protocol
+            #now loop over protocols to add all mappings (needed for chained rule cases)
+            for prot in jsElement['protocols']:
+                #check if prefix is in protocol block
+                if 'prefix' in prot.keys():
+                    #lfn-to-pfn
+                    match = '/(.*)' #match all
+                    result = prot['prefix']+'/$1'
+                    #prefix case should not be chained
+                    chain = None 
+                    rfcInstance.addMapping(str(prot['protocol']), str(match), str(result), chain, 'lfn-to-pfn')
+                    #pfn-to-lfn
+                    match = prot['prefix']+'/(.*)'
+                    result = '/$1'
+                    rfcInstance.addMapping(str(prot['protocol']), str(match), str(result), chain, 'pfn-to-lfn')
+                #here is rules  
+                else:
+                    #loop over rules
+                    for rule in prot['rules']:
+                        match = rule['lfn']
+                        result = rule['pfn']
+                        chain = rule['chain'] if 'chain' in rule.keys() else None
+                        rfcInstance.addMapping(str(prot['protocol']), str(match), str(result), chain, 'lfn-to-pfn')
+                        #pfn-to-lfn: not sure about this!!!
+                        match = rule['pfn'].replace('$1','(.*)')
+                        result = rule['lfn'].replace('/+','/').replace('^/','/')
+                        #now replace anything inside () with $1, for example (.*) --> $1, (store/.*) --> $1
+                        result = re.sub('\(.*\)','$1',result)
+                        rfcInstance.addMapping(str(prot['protocol']), str(match), str(result), chain, 'pfn-to-lfn')
+    
+    return rfcInstance
+
+def rseName(currentSite,currentSubsite,storageSite,volume):
+    """
+    Return Rucio storage element name, for example https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/storage.json?ref_type=heads#L39
+    :currentSite is the site where jobs are executing
+    :currentSubsite is the sub site if jobs are running here
+    :storageSite is the site for storage
+    :volume is the volume name, for example https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/storage.json?ref_type=heads#L3
```

Comment:
Yes, `volume` attribute is needed. As you can see here, we have `WebDAV` protocol in two volumes so `volume` is needed to distinguish them.
https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/storage.json?ref_type=heads#L29
@stlammel, Stephan, could you confirm that `volume` is always present in `<stage-out>` methods? for example:
https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/JobConfig/site-local-config.xml?ref_type=heads#L36
Is there redundancy between `volume` and `rse` in storage description? I mean why we need both terminology which seem to point the same entity?
https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/storage.json?ref_type=heads

Reply:
Duong, based on the storage.json that you shared, I failed to see why `volume` would be relevant for WM stage in/out. For instance:
```
   {  "site": "T1_DE_KIT",
      "volume": "KIT_dCache",
```
or
```
   {  "site": "T1_DE_KIT",
      "volume": "KIT_MSS",
```

Does not contribute to anything in the storage access logic and I would be in favor of removing it from this contribution. However, please let me know if there is anything that I am still missing.

Reply:
Yes, you are right. `volume` does not needed to locate storage.json but it will be needed when we want to identify `rse` (phedex-node) in `rseName` method and read a Rucio File Catalog in `readRFC` method